In [214]:
import math
import time
import requests
import hashlib

# Computational Theory Tasks

## Task 1: Binary Representations

This task involves implementing four functions aimed to manipulate binary data, this is fundemental to many crythographic algorithms. Bitwise Operations allow for the manipulation of individual bits of data, these operations are applied directly to binary representations of numbers.  In cryptography bitwise operations are used to ensure confusion and diffusion in data.  Some common uses of these operations are in Hash functions, Encryption algorithms, Blockchain aswell as many others.

---------------------------------------------------------------------------

### 1. 'rotl(x, n=1)' 
This function rotates bits of a 32-bit unsigned integer 'x' to the left by 'n' places this is effectively shifting bits and wrapping the leftmost bits back into the rightmost position. The most common use for these functions are in crypographic hash functions such as SHA-1 and SHA-256 to provide bitwise diffusion making it difficult to reverse hash.  They are also used in applications like cylic buffers or rotating registers in hardware.

In [215]:
def rotl(x, n=1):
    x &= 0xFFFFFFFF  #ensure x is a 32-bit unsigned integer
    return ((x << n) | (x >> (32 - n))) & 0xFFFFFFFF #rotate left by n bits

#### Step 1 Ensure x is a 32-bit unsigned integer.
- `0xFFFFFFFF` is the hexidecimal constant that represents a 32-bit mask.

- The bitwise AND (&) operation ensures that x only has the lower 32 bits, preventing unexpected large numbers.
#### Step 2 Rotate left by n bits
- `(x << n)` : Left shift moves bits n places to the left, filling with 0s, bits moved beyond 32nd bit are lost

- `(x >> (32 - n))` : Right shift moves the lost bits back to rightmost positions 

- `|` : Bitwise OR combines the shifted values.
### Step 3 Ensure Result is 32-bit
- `& 0xFFFFFFFF` : This ensures result stays within 32-bit unsigned integer range

In [216]:
#example usage:
x = 0b01
n = 3
result = rotl(x, n)
print(f"Original: {x:032b}")
print(f"Rotated Left {n} positions: {result:032b}")

Original: 00000000000000000000000000000001
Rotated Left 3 positions: 00000000000000000000000000001000


---------------------------------------------------------------------------

### 2. 'rotr(x n=1)'
This function rotates bits of a 32-bit unsigned integer 'x' to the right 'n' places this is effectively shifting bits and wrapping the rightmost bits back into the leftmost position. This is similar to the 'rotl' function with the same use cases.

In [217]:
def rotr(x, n=1):
    x &= 0xFFFFFFFF #ensure x is a 32-bit unsigned integer
    return ((x >> n) | (x << (32 - n))) & 0xFFFFFFFF #rotate right by n bits

#### Step 1 Ensure x is a 32-bit unsigned integer.
- `0xFFFFFFFF` is the hexidecimal constant that represents a 32-bit mask.

- The bitwise AND (&) operation ensures that x only has the lower 32 bits, preventing unexpected large numbers.
#### Step 2 Rotate right by n bits
- `(x >> n)` : Right shift moves bits n places to the right, filling with 0s, bits moved beyond 32nd bit are lost

- `(x << (32 - n))` : Left shift moves the lost bits back to leftmost positions 

- `|` : Bitwise OR combines the shifted values.
### Step 3 Ensure result is 32-bit
- `& 0xFFFFFFFF` : This ensures result stays within 32-bit unsigned integer range

In [218]:
#example usage:
x = 0b01
n = 3
result = rotr(x, n)
print(f"Original: {x:032b}")
print(f"Rotated Right by {n} positions: {result:032b}")

Original: 00000000000000000000000000000001
Rotated Right by 3 positions: 00100000000000000000000000000000


---------------------------------------------------------------------------

### 3. 'def ch(x, y, z)'
This function chooses bits from two values: y and z, based off the corresponding bits in x. This works by choosing the corresponding bit in y if the x bit is 1; and choosing the bit corresponding bit in z if x is 0. This is a core compononent in the SHA-2(Secure Hash Algorithm) its used in the message schedule, it works with the bitwise operations to choose part of data from different sources based on values of x. Its also helps mix data from multiple sources to enhace unpredictability of cryptogrpahic transformations.

In [219]:
def ch(x, y, z):
     #choose y if x is true(1), otherwise choose z if x is false(0)
    return (x & y) ^ (~x & z)

#### Step 1 Select y where x is 1
- `(x & y)` : Bitwise AND(&) ensures that only the bits where x is 1 are taken from y
#### Step 2 Select z where x is 0
- `(~x & z) `: Bitwise NOT (~x) inverts bits in x, making the 1s to 0s and vice versa

- Bitwise AND(&) with z ensures only bits were x was originally 0 can be taken form z
#### Step 3 Combine results
- `(x & y) ^ (~x & z)` : Bitwise XOR(^) combines the two selections, ensuring bits from only one (y or z) is taken at each position

In [220]:
#example usage:
x = 0b1100  
y = 0b1010  
z = 0b0110 
result = ch(x, y, z)
print(f"x: {x:04b}, y: {y:04b}, z: {z:04b}")
print(f"ch(x, y, z): {result:04b}")

x: 1100, y: 1010, z: 0110
ch(x, y, z): 1010


---------------------------------------------------------------------------

### 4. 'maj(x, y, z)'
This function takes a majority vote of the bits in each position in x, y and z. Output has 1 in position i if at least two of x, y, z have 1s in position i otherwise the result is 0. This is commonly used in Cryptographic hash functions such as SHA-256 and ensures output is influenced by majority of bits, increasing diffusion, improving resistance to attacks. Its also used in certain Error correction Codes or correction schemes where majority of values are trusted, allowing erros to be flagged based on outliers.

In [221]:
def maj(x, y, z):
    #choose the majority of x, y, z
    return (x & y) ^ (x & z) ^ (y & z)

#### Step 1 Calculate the majority between x and y
- `(x & y)` : Bitwise AND(&) between x and y give the bits where both x and y have 1 in the same position

- This operation shows a majority of the two bits at each position
#### Step 2 Calculate the majority between x and z
- `(x & z)` : Bitwise AND(&) between x and z gives the bits where both x and z have 1 in the same position
#### Step 3 Calculate the majority between y and z
- `(y & z)` : Bitwise AND(&) between y and z gives the bits where both x and z have 1 in the same position
#### Step 4 Combine results with XOR(^)
- `(x & y) ^ (x & z) ^ (y & z)` : Bitwise XOR(^) combines the results of these three majoritys. The XOR operation ensures only the majority value of the three will be choosen

- Eg. If two of the three bits are 1, the result will be 1 fot that bit position


In [222]:
#example usage:
x = 0b1100  
y = 0b1010  
z = 0b0110 
result = maj(x, y, z)
print(f"x: {x:04b}, y: {y:04b}, z: {z:04b}")
print(f"maj(x, y, z): {result:04b}")

x: 1100, y: 1010, z: 0110
maj(x, y, z): 1110


---------------------------------------------------------------------------

### Test function
Testing each Binary operation function using hexdecimal number representations

In [223]:
#test cases
def test_binary_functions():
    #test rotl using hexadecimal numbers
    #0x01 = 0000 0001, 0x02 = 0000 0010
    assert rotl(0x01, 1) == 0x02
    #0x80000000 = 1000 0000 0000 0000 0000 0000 0000 0000, 0x00000001 = 0000 0000 0000 0000 0000 0000 0000 0001
    assert rotl(0x80000000, 1) == 0x00000001
    #0x12345678 = 0001 0010 0011 0100 0101 0110 0111 1000, 0x23456781 = 0010 0011 0100 0101 0110 0111 1000 0001 
    assert rotl(0x12345678, 4) == 0x23456781

    #test rotr using hexadecimal numbers
    #0x01 = 0000 0001, 0x02 = 0000 0010
    assert rotr(0x02, 1) == 0x01
    #0x80000000 = 1000 0000 0000 0000 0000 0000 0000 0000, 0x00000001 = 0000 0000 0000 0000 0000
    assert rotr(0x00000001, 1) == 0x80000000
    #0x12345678 = 0001 0010 0011 0100 0101 0110 0111 1000, 0x81234567 = 1000 0001 0010 0011 0100 0101 0110 0111
    assert rotr(0x12345678, 4) == 0x81234567

    #test ch using hexadecimal numbers
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0xAAAAAAAA = 1010 1010 1010 1010 1010 1010 1010 1010, 0x55555555 = 0101 0101 0101 0101 0101 0101 0101 0101
    assert ch(0xFFFFFFFF, 0xAAAAAAAA, 0x55555555) == 0xAAAAAAAA #x is all ones, so the result is y
    #0x00000000 = 0000 0000 0000 0000 0000 0000 0000 0000, 0xAAAAAAAA = 1010 1010 1010 1010 1010 1010 1010 1010, 0x55555555 = 0101 0101 0101 0101 0101 0101 0101 0101
    assert ch(0x00000000, 0xAAAAAAAA, 0x55555555) == 0x55555555 #x is all zeros, so the result is z
    #0xF0F0F0F0 = 1111 0000 1111 0000 1111 0000 1111 0000, 0xAAAAAAAA = 1010 1010 1010 1010 1010 1010 1010 1010, 0x55555555 = 0101 0101 0101 0101 0101 0101 0101 0101
    assert ch(0xF0F0F0F0, 0xAAAAAAAA, 0x55555555) == 0xA5A5A5A5 #0xA5A5A5A5 = 1010 0101 1010 0101 1010 0101 1010 0101 (combination of y and z as x is mixed)

    #test maj using hexadecimal numbers
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0x00000000 = 0000 0000 0000 0000 0000 0000 0000 0000
    assert maj(0xFFFFFFFF, 0x00000000, 0x00000000) == 0x00000000 # majority is 0x00000000 so the result is 0x00000000
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0x00000000 = 0000 0000 0000 0000 0000 0000 0000 0000
    assert maj(0xFFFFFFFF, 0xFFFFFFFF, 0x00000000) == 0xFFFFFFFF # majority is 0xFFFFFFFF so the result is 0xFFFFFFFF
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0xF0F0F0F0 = 1111 0000 1111 0000 1111 0000 1111 0000
    assert maj(0xFFFFFFFF, 0xF0F0F0F0, 0xF0F0F0F0) == 0xF0F0F0F0 # majority is 0xF0F0F0F0 so the result is 0xF0F0F0F0

    print("All test cases passed successfully")

test_binary_functions()

All test cases passed successfully


---------------------------------------------------------------------------

### Task 1 References:
1. FIPS PUB 180-4: Secure Hash Standard (SHS) :https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf  
2. M. A. Hossain, M. M. Alam, and M. S. Alam, "Cryptography Encryption Technique Using Circular Bit Rotation in Binary Field," 2020 IEEE Region 10 Symposium (TENSYMP), Dhaka, Bangladesh, 2020: https://ieeexplore.ieee.org/document/9197845/
3. Rotate bits of a number: https://www.geeksforgeeks.org/rotate-bits-of-an-integer/
4. RealPython: Bitwise operators in python: https://realpython.com/python-bitwise-operators/  
5. Understanding Bitwise Operations in Python: Cryptography, Hashing, and Real-World Applications: https://www.linkedin.com/pulse/understanding-bitwise-operations-python-cryptography-hashing-woon-sm7ec/

---------------------------------------------------------------------------

## Task 2: Hash Functions

### Polynomial Rolling Hash Function
Hashing is common practice in the world of computer science used to map data, such as strings, to numerical values. A widely used approach for hasing strings is the polynomial rolling hash function.
This function is defined as:  
$$H(s) = (s_1 \times p^{(n-1)} + s_2 \times p^{(n-2)} + ... + s_n \times p^0) \mod m$$

- $H(s)$ is hash value of the string
- $s1, s2, sn$ are ASCII values of characters in the string
- $p$ is a prime base (commonly 31)
- $m$ is the modulus value (commonly 101)
- $n$ is the lenght of the string 

The Polynomial Rolling Hash Function has many use cases and is widely used in
- String Matching Algorithms
- Data Structures
- Plagiarism Detection
- Crypography

In [224]:
def hash(s):
    #initialise the hash function to 0
    hashval = 0

    #iterate over each character in the string
    for c in s:
        #compute hash using polynomial rolling hash function
        hashval = ord(c) + 31 * hashval
        print(f"Character: {c}, Hash: {hashval}")
    
    #apply modulo operation to keep hash within fixed range (0-100)
    hashval =  hashval % 101

    return f"Final Hash Value: {hashval}"

### Code Explanation

- This variable will store the computed hash value, this will be happen as the function processes each charecter in the string.

In [225]:
hashval = 0

### Polynomial Rolling Hash Calculation
- This itereates over each charecter `c` in the string `s`. Each character in `s` contributes to the final hash.

- `ord(c)`: Converts the character `c` into ASCII value.

- `31 * hashval` : This takes the current hash and multiplies it by 31, this is a prime number commonly used for hash functions.

- `ord(c) + 31 * hashval` : The full line adds the ASCII of the character to a scaled hash value, this is to ensure each character incluences the hash in differently depending on the characters position.

In [226]:
#string to be hashed
s = "hello"

#interate over each character in the hello
for c in s:
    #compute hash using polynomial rolling hash function
    hashval = ord(c) + 31 * hashval

print(f"Hash Value: {hashval}")

Hash Value: 99162322


### Modulo Operation (%101)
- This aims to ensure the final hash is in a specified range (0-100)

- This is to avoid faults like integer overflow, but also keeps the hash value compact.

- The prime modulous (101) is used as it helps distribute hash values evenly, which helps reduce collisions.

In [227]:
hashval % 101

17

### Example of how function works:

In [228]:
print(hash("hello"))

Character: h, Hash: 104
Character: e, Hash: 3325
Character: l, Hash: 103183
Character: l, Hash: 3198781
Character: o, Hash: 99162322
Final Hash Value: 17


Iterate over each charecter in hello applying rolling hash calculation:
1. h (ASCII 104): hashval = 104
2. e (ASCII 101): hashval = 101 + 31 x 104 = 3325
3. l (ASCII 108): hashval = 108 + 31 x 3325 = 103183
4. l (ASCII 108): hashval = 108 + 31 x 103183 = 3198781
5. o (ASCII 111): hashval = 111 + 31 x 3198781 = 99162322

Apply modulo operation:
99162322 mod 101 = 17

"Hello" = 17

### Why use 31 as the Base?
1. Efficient Computation:
    - 31 is close to the powers of 2, this allows efficient multiplication and bitwise operations on many processors.
2. Empirical Evidence:
    - Many real-world applications use 31 due to it providing a good balance of speed and low collision probability.
3. Prime Number Property:
    - Prime numbers reduce chances of hash collisions occuring this is done by ensuring more uniform distribution of hash values

### Why use 101 as the Modulus?
1. Prime Numbers Improve Hash Distribution:
    - Prime modulus ensures values are distributed evenly this inturn reduces hash collisions
2. Keeps Hash Values Compact:
    - Modulo operations keep hash values in a range in this case that range is 0 -100 this makes it more efficient for hashing tables.

---------------------------------------------------------------------------

### Task 2 References:
1. String hasing using Polynomial rolling hash function: https://www.geeksforgeeks.org/string-hashing-using-polynomial-rolling-hash-function
2. String hashing: https://cp-algorithms.com/string/string-hashing.html
3. Why should hash functions use a prime number modulus?: https://www.designgurus.io/answers/detail/why-should-hash-functions-use-a-prime-number-modulus
4. Understanding Rolling Hash: A Key Component in String Matching Algorithms: https://medium.com/%40ggaappuu1234/understanding-rolling-hash-a-key-component-in-string-matching-algorithms-83236d8c4a20
5. Introduction to Rolling Hash – Data Structures and Algorithms: https://www.geeksforgeeks.org/introduction-to-rolling-hash-data-structures-and-algorithms/
6. On the mathematics behind rolling hashes and anti-hash tests: https://codeforces.com/blog/entry/60442

---------------------------------------------------------------------------

## Task 3: SHA256

SHA-256 is a cryptographic hash function that produces a 256 bit hash value, its part of the SHA-2 family designed by NSA and used as a federal standard(FIPS-180-4).

### Why Padding is Necessary
SHA-256 processes messages in 512 bit blocks. Messages must be padded to:
- Create complete 512 bit blocks
- Include the orignal message lenght
- Ensure the hash is computationally infeasible to reverse

In [229]:
def sha256(file_path):
    #read the file contents
    with open(file_path, 'r') as f:
        binary_content = f.read().replace(' ', '')
    
    #convert binary string to bytes
    data = bytes(int(binary_content[i:i+8], 2) for i in range(0, len(binary_content), 8))
    
    #calculate original length in bits
    orig_len = len(data) * 8
    
    #start padding by adding the 1 bit (0x80 in hex)
    pad_data = data + b'\x80'
    
    #calculate padding length to make total length
    pad_len = (56 - (len(pad_data) % 64)) % 64
    
    #add zero padding
    pad_data += b'\x00' * pad_len
    
    #append original length as 64-bit big-endian unsigned integer
    pad_data += orig_len.to_bytes(8, 'big')
    
    #print padding in hex
    padding_hex = " ".join(f"{byte:02x}" for byte in pad_data[len(data):])
    print(padding_hex)
    

In [230]:
file_path = "test.txt"
sha256(file_path)

80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 18


### SHA-256 Padding Specification (FIPS-180-4)
The padding process consists of 3 steps:
#### Step 1: Append '1' bit
add a single '1' bit to the end of the message
- in byte format: 0x80 (100000000 in binary)
#### Step 2: Append '0' bits
Add enough 0 bits so the total lenght is congruent to 449 mod 512
- This leaves exactly 64 bits for the lenght information
- Formula: (448-(current_lenght + 1)) mod 512
#### Step 3: Append lenght
Add the orignal message lenght as a 64 bit big endian integer
- Represents the lenght in bits
- Allows messages up to 2^64 - 1 bits long

### Code explanation

In [231]:
# Step 1: Read and Process File
print("Step 1: Reading file contents")
#read the file contents
with open(file_path, 'r') as f:
    original_content = f.read()
    binary_content = original_content.replace(' ', '')

#convert to binary contents 
print (f"Original Content: {original_content}")
print (f"Binary Content: {binary_content}")

Step 1: Reading file contents
Original Content: 01100001 01100010 01100011
Binary Content: 011000010110001001100011


In [232]:
# Step 2: Convert Binary String to Bytes
print("Step 2: Converting binary string to bytes")
data = bytes(int(binary_content[i:i+8], 2) for i in range(0, len(binary_content), 8))

for i in range(0, len(binary_content), 8):
    binary_chunk = binary_content[i:i+8]
    byte_value = int(binary_chunk, 2)
    print(f"Binary: {binary_chunk} -> Decimal: {byte_value} -> Hex: {byte_value:02x}")

Step 2: Converting binary string to bytes
Binary: 01100001 -> Decimal: 97 -> Hex: 61
Binary: 01100010 -> Decimal: 98 -> Hex: 62
Binary: 01100011 -> Decimal: 99 -> Hex: 63


In [233]:
# Step 3: Calculate Original Length
print("Step 3: Calculating original length")
orig_len = len(data) * 8
print(f"Data length: {len(data)} bytes = {orig_len} bits\n")

Step 3: Calculating original length
Data length: 3 bytes = 24 bits



In [234]:
# Step 4: Add '1 Bit
print("Step 4: Adding '1' bit (0x80)")
pad_data = data + b'\x80'
print(f"0x80 in binary: {0x80:08b}")
print(f"Data after adding 0x80: {pad_data}\n")

Step 4: Adding '1' bit (0x80)
0x80 in binary: 10000000
Data after adding 0x80: b'abc\x80'



In [235]:
# Step 5: Calculate Zero Padding
print("Step 5: Calculating zero padding length")
pad_len = (56 - (len(pad_data) % 64)) % 64
print(f"Current length: {len(pad_data)} bytes")
print(f"Target length: 56 bytes (448 bits)")
print(f"Padding needed: {pad_len} bytes\n")

Step 5: Calculating zero padding length
Current length: 4 bytes
Target length: 56 bytes (448 bits)
Padding needed: 52 bytes



In [236]:
# Step 6: Add Zero Padding
print("Step 6: Adding zero padding")
pad_data += b'\x00' * pad_len
print(f"Data after zero padding: length = {len(pad_data)} bytes\n")

Step 6: Adding zero padding
Data after zero padding: length = 56 bytes



In [237]:
# Step 7: Add Length
print("Step 7: Adding original length as 64-bit big-endian")
length_bytes = orig_len.to_bytes(8, 'big')
print(f"Original length: {orig_len} bits")
print(f"As 64-bit big-endian: {' '.join(f'{b:02x}' for b in length_bytes)}")
pad_data += length_bytes
print(f"Final padded data length: {len(pad_data)} bytes\n")

Step 7: Adding original length as 64-bit big-endian
Original length: 24 bits
As 64-bit big-endian: 00 00 00 00 00 00 00 18
Final padded data length: 64 bytes



In [238]:
# Step 8: Print Final Padding
print("Step 8: Final padding in hex")
padding_hex = " ".join(f"{byte:02x}" for byte in pad_data[len(data):])
print(padding_hex)

Step 8: Final padding in hex
80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 18


### Example Analysis
For the message "abc" (binary: 01100001 01100010 01100011)
1. Orignal: 3 bytes = 24 bits
2. After '1' bit: 24 + 1 = 25 bits
3. Padding required: 488 - 25 = 423 bits
4. Zero padding: 423 bits = 52 bytes
5. Lenght to append: 8 bytes
6. Total: 3 + 1 + 52 + 8 = 64 bytes

Final padding is: 80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 18

---------------------------------------------------------------------------

### Task 3 References:
1. FIPS PUB 180-4: Secure Hash Standard (SHS): https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf  
2. Schneier, B.(2015). Applied Cryptography: https://www.schneier.com/books/applied-cryptography/
3. Menezes, A. et al.(1996) Handbook of Applied Cryptography: https://galois.azc.uam.mx/mate/propaganda/Menezes.pdf
4. NIST Special Publication 800-107: Recommendation for Applications Using Approved Hash Algorithms

---------------------------------------------------------------------------

## Task 4: Prime Numbers

### Algoritm 1: Sieve of Eratosthenes

### How it works
Find the first n prime numbers using the Sieve of Eratosthenes algorithm.

Algorithm Explanation
1. Create a boolean array for numbers from 2 to an estimated limit
2. Mark all multiples of each prime starting from 2
3. All unmarked numbers at the end are prime
4. Extract the first n prime numbers from the sieve

Advantages: 
- Very efficient for finding many primes in a range
- Significantly faster that trial division for large n
- Classic algorithm known for it efficiency

Disadvantages:
- Requires more memory than trial division
- Need to estimate upper limit for n primes

In [239]:
def sieve_of_eratosthenes(n):
    #estimate upper limit for n prime number
    #using prime number theorem approximation
    if n < 6:
        limit = 15
    else:
        limit = int(n * (math.log(n) + math.log(math.log(n))))

    #create sieve array - true means prime
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False #0 and 1 are not prime

    #mark multiples of each prime
    for i in range(2, int(math.sqrt(limit)) + 1):
        if sieve[i]:
            #mark all multiples of i as not prime
            for j in range(i * i, limit + 1, i):
                sieve[j] = False

    #collect prime numbers
    primes = []
    current = 2
    while len(primes) < n and current <= limit:
        if sieve[current]:
            primes.append(current)
        current += 1
    return primes

In [240]:
print("Finding first 100 prime numbers using Sieve of Eratosthenes")
start_time = time.time()
n = 100
primes = sieve_of_eratosthenes(n)
print(f"First {n} prime numbers: {primes} \nin {time.time() - start_time:.4f} seconds")

Finding first 100 prime numbers using Sieve of Eratosthenes
First 100 prime numbers: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541] 
in 0.0003 seconds


---------------------------------------------------------------------------

### Algorithm 2: Trial Division Method

### How it works
Find the first n prime numbers using trial division algorithm

Algorithm Explanation:
1. Start checking from 2 (first prime)
2. For each number check if its divisible by any number from 2 to sqrt(n)
3. If no divisors found its a prime
4. Continue until we find n primes

Advantages:
- Simple to implement
- Small memory requirment
- Can find primes one by one as needed

Disadvantages:
- Slower that other algorithms for large n
- Redundant calculations (checks same number multiple times)

In [241]:
def trial_division(n):
    primes = []
    current = 2
    while len(primes) < n:
        is_prime = True
        #optimise by only checking to square root of current num
        for i in range(2, int(math.sqrt(current))+1):
            if current % i == 0:
                is_prime = False
                break
        if is_prime:
            primes.append(current)
        #optimise after 2 only check odd numbers
        current += 1 if current == 2 else 2

    return primes

In [242]:
print("Finding first 100 prime numbers using trial division")
start_time = time.time()
n = 100
primes = trial_division(n)
print(f"First {n} prime numbers: {primes} \nin {time.time() - start_time:.4f} seconds")

Finding first 100 prime numbers using trial division
First 100 prime numbers: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541] 
in 0.0006 seconds


---------------------------------------------------------------------------

1. Sieve of Eratosthenes: https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes
2. Prime Number in Python: https://www.ccbp.in/blog/articles/prime-number-in-python
3. Trial division(article) Khan academy : https://www.khanacademy.org/computing/computer-science/cryptography/comp-number-theory/a/trial-division
4. Fastest Algoritm to Find Prime Numbers, Baeldung: https://www.baeldung.com/cs/prime-number-algorithms
5. Check for Prime Number, GeeksforGeeks: https://www.geeksforgeeks.org/check-for-prime-number/


---------------------------------------------------------------------------

## Task 5: Roots

In [243]:
def fractional_bits_of_sqrt(n, bits=32):
    #calculate square root
    sqrt = math.sqrt(n)
    
    #extract fractional part
    integer = int(sqrt)
    fraction = sqrt - integer

    #convert to binary
    binary_bits = []
    current_fraction = fraction
    for _ in range(bits):
        #multiply by 2
        current_fraction *= 2
        #get integer part
        bit = int(current_fraction)
        binary_bits.append(str(bit))
        #keep only the fractional part
        current_fraction -= bit

    return ''.join(binary_bits)
        

In [244]:
def get_bits(primes):
    results = []
    #iterate over each prime number
    for prime in primes:
        #get 32 bits of fractional part
        binary_fraction = fractional_bits_of_sqrt(prime, 32)

        results.append({
            'prime': prime,
            'sqrt_fraction': binary_fraction,
        })
    return results

In [245]:
primes = sieve_of_eratosthenes(100)
sqrt_bits = get_bits(primes)
print("Fractional bits of square root of first 100 prime numbers:")
for result in sqrt_bits:
    prime = result['prime']
    sqrt_fraction = result['sqrt_fraction']
    print(f"Prime: {prime}, Fractional bits of sqrt: {sqrt_fraction}")

Fractional bits of square root of first 100 prime numbers:
Prime: 2, Fractional bits of sqrt: 01101010000010011110011001100111
Prime: 3, Fractional bits of sqrt: 10111011011001111010111010000101
Prime: 5, Fractional bits of sqrt: 00111100011011101111001101110010
Prime: 7, Fractional bits of sqrt: 10100101010011111111010100111010
Prime: 11, Fractional bits of sqrt: 01010001000011100101001001111111
Prime: 13, Fractional bits of sqrt: 10011011000001010110100010001100
Prime: 17, Fractional bits of sqrt: 00011111100000111101100110101011
Prime: 19, Fractional bits of sqrt: 01011011111000001100110100011001
Prime: 23, Fractional bits of sqrt: 11001011101110111001110101011101
Prime: 29, Fractional bits of sqrt: 01100010100110100010100100101010
Prime: 31, Fractional bits of sqrt: 10010001010110010000000101011010
Prime: 37, Fractional bits of sqrt: 00010101001011111110110011011000
Prime: 41, Fractional bits of sqrt: 01100111001100110010011001100111
Prime: 43, Fractional bits of sqrt: 100011101011

---------------------------------------------------------------------------

## Task 6: Proof of Work

In [246]:
def count_leading_zeros(word):
    #get the SHA-256 hash of the word
    hash_object = hashlib.sha256(word.encode('utf-8'))
    hash_digest = hash_object.digest()

    #count leading zeros
    leading_zeros = 0
    for byte in hash_digest:
        #process each byte
        for bit_position in range(8):
            #check if the bit is 0
            if not (byte & (128 >> bit_position)):
                leading_zeros += 1
            else:
                #stop counting at first 1
                return leading_zeros
    #return total leading zeros
    return leading_zeros

In [247]:
def most_leading_zeros(n=1):
    #set the dictionary file path
    dictionary_file = 'words.txt'
    #open dictionary file
    with open(dictionary_file, 'r') as f:
        words = [line.strip() for line in f if line.strip()]

    word_zeros = []

    #iterate over each word
    for word in words:
        #make all words lowercase has more leading zeros
        #word = word.lower()
        #count leading zeros
        zeros = count_leading_zeros(word)
        word_zeros.append((word, zeros))

    #sort by leading zeros
    word_zeros.sort(key=lambda x: x[1], reverse=True)

    return word_zeros[:n]

In [248]:
top_10_words = most_leading_zeros(10)
print("Top 10 words with most leading zeros:")
for word, zeros in top_10_words:
    print(f"Word: {word}, Leading Zeros: {zeros}")

Top 10 words with most leading zeros:
Word: APPLICANT, Leading Zeros: 16
Word: ANACHRONISTICALLY, Leading Zeros: 15
Word: NECK, Leading Zeros: 14
Word: DRYLY, Leading Zeros: 13
Word: METAMORPHOSIS, Leading Zeros: 13
Word: PLATOON, Leading Zeros: 13
Word: SURGERY, Leading Zeros: 13
Word: BERESFORD, Leading Zeros: 12
Word: INTERFACED, Leading Zeros: 12
Word: POISONS, Leading Zeros: 12


---------------------------------------------------------------------------

## Task 7: Turing Machines

---------------------------------------------------------------------------

## Task 8: Computational Complexity

---------------------------------------------------------------------------

## End